### PR TITLE
Onboard CodeVersion and CodeContainer

### PR DESCRIPTION
**Pull request by forteddyt**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+			},
+			"additionalProperties": false
+		},
+		"CodeVersion": {
+			"description": "Details of an RegisteredCode",
+			"type": "object",
+			"properties": {
+				"root": {
+					"$ref": "#/definitions/DirTreeNode"
+				},
```
Comment (j-so) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:393]: You can remove this (and remove the definition for DirTreeNode.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+		"https"
+	],
+	"consumes": [
+		"application/json"
+	],
+	"produces": [
+		"application/json"
+	],
+	"paths": {
+		"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/workspaces/{workspaceName}/Codes/codeversions": {
```
Comment (j-so) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:19]: TThis should end with "/codes". You can remove the "codeversions" part (bug).

Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:19]: should I lowercase all `/Codes` to `/codes`?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+      "delete": {
+        "tags": [
+          "CodeContainer"
+        ],
+        "x-ms-examples": {
+          "Delete CodeContainer": {
+            "$ref": "./examples/deleteCodeContainer.json"
+          }
+        },
+        "description": "Delete CodeVersion entity.",
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:95]: are these containers or versions?

Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:95]: Are these containers or versions?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+          }
+        ],
+        "responses": {
+          "default": {
+            "description": "Error",
+            "schema": {
+              "$ref": "machineLearningServices.json#/definitions/MachineLearningServiceError"
+            }
+          },
+          "200": {
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:121]: delete should return 204 if the resource did not exist

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+      "description": "Name for CodeVersion.",
+      "required": true,
+      "x-ms-parameter-location": "method"
+    }
+  },
+  "definitions": {
+    "AssetContainer": {
+      "type": "object",
+      "properties": {
+        "name": {
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:242]: name is a top level field in AssetContainerResource.  It shouldn't be duplicated here.  If this is some sort of other display name please change the property name to make it more clear

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+          "type": "string"
+        },
+        "isDirectory": {
+          "description": "Whether the path defines a directory or a single file.",
+          "type": "boolean"
+        }
+      },
+      "additionalProperties": false
+    },
+    "CodeVersion": {
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:311]: Please remove the definitions that aren't used in this file

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+            "$ref": "machineLearningServices.json#/parameters/WorkspaceNameParameter"
+          },
+          {
+            "$ref": "machineLearningServices.json#/parameters/APIVersionParameter"
+          },
+          {
+            "$ref": "machineLearningServices.json#/parameters/PaginationParameter"
+          },
+          {
+            "$ref": "#/parameters/CodeVersionNameParameter"
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:69]: this shouldn't take a name parameter

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+      "in": "path",
+      "name": "version",
+      "type": "string",
+      "description": "Version for CodeVersion.",
+      "required": true,
+      "x-ms-parameter-location": "method"
+    },
+    "CodeVersionVersionOptionalParameter": {
+      "in": "query",
+      "name": "version",
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:265]: what is this query parameter used for?  If the user wants a specific version they can do an individual GET on that version

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+    "CodeVersionResourcePaginatedResult": {
+      "type": "object",
+      "properties": {
+        "value": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/CodeVersionResource"
+          }
+        },
+        "continuationToken": {
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:380]: the collection GET contract should only contain "value" and "nextLink"

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+          "type": "boolean"
+        }
+      },
+      "additionalProperties": false
+    },
+    "CodeVersion": {
+      "description": "Details of an RegisteredCode",
+      "type": "object",
+      "properties": {
+        "name": {
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:291]: remove or rename.  There is a "name" property at the top level which should indicate the resources name.  The name of the code container is also already present in the URI

**Review by pilor — CHANGES_REQUESTED** [review]

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json (hunk 6 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:311]: which definitions aren't being used in this file? I might be missing something, but it looks like they're all being used:
AssetContainer is in AssetContainerResource [in Code_Get + others] is in AssetContainerResourcePaginatedResult [in CodeVersions_List]
AssetPath is in CodeVersion is in CodeVersionResourceRequest [in Code_CreateOrUpdate]

Comment (pilor) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:311]: CodeVersion definitions should go in the codeversion file.  I think these are duplicates to what is in there

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json (hunk 3 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:95]: fixed
please reflect changes in controller @j-so

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json (hunk 4 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:121]: fixed
please reflect changes in controller @j-so

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json (hunk 8 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:265]: you're right, that shouldn't exist
please reflect changes in controller @j-so

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json (hunk 6 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json]: good catch - fixed :)

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+            "$ref": "#/parameters/CodeVersionVersionParameter"
+          },
+          {
+            "in": "body",
+            "name": "body",
+            "description": "Code entity to create or update.",
+            "schema": {
+              "$ref": "#/definitions/CodeVersionResourceRequest"
+            }
+          }
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:218]: Why is a codeversion API returning a codecontainer resource?

**Review by pilor — CHANGES_REQUESTED** [review]
1 new comment + existing comments from previous review

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json (hunk 5 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json]: should these sort of duplicates show up in the top-most level field (ie AssetContainerResource) or the bottom-most level field (ie AssetContainer)?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json (hunk 11 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:218]: whoops, it shouldn't :D fixed

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json (hunk 9 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:373]: fixed

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+          "description": "The asset tag dictionary. Tags can be added, removed, and updated."
+        },
+        "latestVersions": {
+          "type": "array",
+          "items": {
+            "$ref": "codeVersion.json#/definitions/CodeVersionResource"
+          },
+          "description": "Latest Versions for all assets"
+        },
+        "name": {
```
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:306]: The inner "name" property on both CodeContainer and CodeVersion needs to be removed or renamed.  You shouldn't re-use top level property names in your property bag

**Review by pilor — CHANGES_REQUESTED** [review]

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json (hunk 12 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:306]: should these sort of duplicates show up in the top-most level field or the bottom-most level field?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json (hunk 10 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:283]: should these sort of duplicates show up in the top-most level field or the bottom-most level field?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json (hunk 12 above)
Comment (pilor) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json:306]: "name" is a required top level property on every resource.  It should only exist there

Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeContainer.json]: some "name" properties were removed. The other name properties belong, but were not labeled as an azure-resource correctly. Everything _should_ be good now :)

**Review by pilor — APPROVED** [review]

**Comment by forteddyt** [comment]
how do I locally run this LintDiff gate? I'm assuming it's [this part of the docs](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/ci-fix.md#breaking-change-check), but I'm confused on what to compare against since this is a new spec

**Comment by ArcturusZhang** [comment]
Hi @forteddyt [here](https://github.com/Azure/azure-rest-api-specs/blob/master/.azure-pipelines/LintDiff.yml) is the scripts running in CI, you could follow those steps to make it run locally

**Comment by forteddyt** [comment]
I must be missing something. Running
```
npm install
scripts/install-dotnet.sh
npm run tsc && node scripts/momentOfTruth.js && node scripts/momentOfTruthPostProcessing.js
```
gives me back  `0 new errors / 0 new warnings`

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/workspaces/{workspaceName}/codes/{name}/versions": {
+      "get": {
+        "tags": [
+          "CodeVersion"
+        ],
+        "x-ms-pageable": {
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/codeVersion.json:42]: This is correct

**Review by ArcturusZhang — APPROVED** [review]
Please let me know when this is ready to merge

**Pull request closed without merging**